rasa_core/channels/botframework.py

```diff
@@ -0,0 +1,169 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import json
+import logging
+from requests import post
+import datetime
+
+from flask import Blueprint, request, jsonify
+
+from rasa_core.channels.channel import UserMessage, OutputChannel
+from rasa_core.channels.rest import HttpInputComponent
+
+logger = logging.getLogger(__name__)
+
+
+class BotFramework(OutputChannel):
+    """A Microsoft Bot Framework communication channel."""
+
+    token_expiration_date = datetime.datetime.now()
+    headers = None
+
+    def __init__(self, bf_id, bf_secret, conversation, bot_id,
+                 service_url):
+        # type: (Text, Text, Dict[Text]) -> None
+
+        self.bf_id = bf_id
+        self.bf_secret = bf_secret
+        self.conversation = conversation
+        self.global_uri = "%sv3/" % service_url
+        self.bot_id = bot_id
+
+    def get_headers(self):
+        if BotFramework.token_expiration_date < datetime.datetime.now():
+            microsoft_oauth2_url = 'https://login.microsoftonline.com'
+            microsoft_oauth2_path = 'botframework.com/oauth2/v2.0/token'
+            uri = "%s/%s" % (microsoft_oauth2_url, microsoft_oauth2_path)
+            grant_type = 'client_credentials'
+            scope = 'https://api.botframework.com/.default'
+            payload = {'client_id': self.bf_id,
+                       'client_secret': self.bf_secret,
+                       'grant_type': grant_type,
+                       'scope': scope}
+
+            token_response = post(uri, data=payload)
```

Comment:
how does this work within multiple processes, so e.g. if there are multiple instances of this class in separate processes, can there be multiple active `access_tokens` or is it always just one?

Reply:
`token_expiration_date` and `headers` variables are class-level variables which are independent from any instance-level variable. This way, someone can deploy a bot on various `BotFramework`.

Reply:
thanks, but I am more concerned about the validity of the access token: What happens if there are two instances of this class, both with different access_tokens, can they both be valid or is there always only one acctive token (don't know the botframework to well).